src/servers/FtpServer.cc

```diff
@@ -1439,7 +1439,25 @@ Ftp::Server::createDataConnection(Ip::Address cltAddr)
     Comm::ConnectionPointer conn = new Comm::Connection();
     conn->flags |= COMM_DOBIND;
 
-    if (clientConnection->flags & COMM_INTERCEPTION) {
+    if (port->flags.tproxyInterceptLocally() || port->flags.explicitProxy()) {
```

Comment:
Behaviour change. Please undo the "polish". This should be sufficient here:
```
    if (port->flags.natInterceptLocally()) {
    ...
```

FYI;
The original code was specifically designed for the default behaviour to be sharing of IP address between FTP CTRL and FTP DATA connections. The re-ordering of these if-else clauses inverts that default behaviour to make the common case the one where client is told to connect with Squid machines primary IP:port. That causes unnecessary problems for clients whose basic security rejects DATA connection IP being different from CTRL.

Reply:
>  This should be sufficient here:

I do not mind swapping the if/else checks order as you suggest. Sorry, but I could not understand the behavior change you were talking about - i.e., why it should matter which one of the mutual exclusive conditions we check first.